docs/groovy-script/mods/abyssaltweaker/ritual.md

```diff
@@ -0,0 +1,158 @@
+
+:::::::::: details Warning {open id="warning"}
+For technical reasons the Ritual compatibility will not validate whether the ritual is executed in a valid dimension. By default only Overworld and Abyssalcraft's dimensions are valid, and other dimensions have to be added through the config file.
+::::::::::
+
+## Identifier
+
+Refer to this via any of the following:
+
+```groovy:no-line-numbers {1}
+mods.abyssaltweaker.ritual/* Used as page default */ // [!code focus]
+mods.abyssaltweaker.Ritual
+```
+
+
+## Adding Recipes
+
+### Recipe Builder
+
+Just like other recipe types, the Necronomicon Ritual also uses a recipe builder.
+
+Don't know what a builder is? Check [the builder info page](../../getting_started/builder.md) out.
+
+:::::::::: details mods.abyssaltweaker.ritual.recipeBuilder() {open id="abstract"}
+- `IngredientList<IIngredient>`. Sets the item inputs of the recipe. Requires greater than or equal to 1 and less than or equal to 9.
+
+    ```groovy:no-line-numbers
+    input(IIngredient)
+    input(IIngredient...)
+    input(Collection<IIngredient>)
+    ```
+
+- `ItemStackList`. Sets the item outputs of the recipe. Requires exactly 1.
+
+    ```groovy:no-line-numbers
+    output(ItemStack)
+    output(ItemStack...)
+    output(Collection<ItemStack>)
+    ```
+
+- `int`. Sets the amount of Potential Energy the ritual consumes. Requires greater than or equal to 1. (Default `0`).
+
+    ```groovy:no-line-numbers
+    pe(int)
+    energy(int)
+    ```
+
+- `String`. Sets the ritual name, which has to be unique among all rituals. Requires not empty.
+
+    ```groovy:no-line-numbers
+    name(String)
+    name(ResourceLocation)
+    ```
+
+- `int`. Sets the minimum book tier that can execute this recipe. Requires greater than or equal to 0 and less than or equal to 4. (Default `0`).
+
+    ```groovy:no-line-numbers
+    bookTier(int)
+    bookTier(String)
+    ```
+
+- `int`. Sets the particle ID that is emitted from pedestals while this recipe is being executed (see also: EnumRitualParticle). Requires greater than or equal to 0 and less than 8. (Default `3`).
+
+    ```groovy:no-line-numbers
+    particle(int)
+    ```
+
+- `int`. Sets the dimension that this ritual can be performed in. (Default `any dimension`).
+
+    ```groovy:no-line-numbers
+    dimension(int)
+    ```
+
+- `boolean`. Sets whether the ritual requires a sacrifice. (Default `false`).
+
+    ```groovy:no-line-numbers
+    sacrifice(boolean)
+    requiresSacrifice()
+    ```
+
+- First validates the builder, returning `null` and outputting errors to the log file if the validation failed, then registers the builder and returns the registered object. (returns `null` or `com.shinoow.abyssalcraft.api.ritual.NecronomiconInfusionRitual`).
+
+    ```groovy:no-line-numbers
+    register()
+    ```
+
+::::::::: details Example {open id="example"}
+```groovy:no-line-numbers
+mods.abyssaltweaker.ritual.recipeBuilder()
```

Comment:
can you add more examples of the possible ways to write this builder?

Reply:
like one with the least parameters possible and another one with the most parameters possible or something?

Reply:
yeah, minimum + maximum, and if there are distinct options - eg if entity rituals exist - a pair for those too.